services/motion/builtin/state/state.go

```diff
@@ -0,0 +1,619 @@
+	}
+	if err := e.state.updateStateNewPlan(msg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *execution) notifyStateNewPlan(plan motion.Plan, time time.Time) error {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+	return e.state.updateStateNewPlan(newPlanMsg{
+		plan:       plan,
+		planStatus: motion.PlanStatus{State: motion.PlanStateInProgress, Timestamp: time},
+	})
+}
+
+func (e *execution) notifyStatePlanFailed(plan motion.Plan, reason string, time time.Time) error {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+	return e.state.updateStateStatusUpdate(stateUpdateMsg{
+		componentName: e.componentName,
+		executionID:   e.id,
+		planID:        plan.ID,
+		planStatus:    motion.PlanStatus{State: motion.PlanStateFailed, Timestamp: time, Reason: &reason},
+	})
+}
+
+func (e *execution) notifyStatePlanSucceeded(plan motion.Plan, time time.Time) error {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+	return e.state.updateStateStatusUpdate(stateUpdateMsg{
+		componentName: e.componentName,
+		executionID:   e.id,
+		planID:        plan.ID,
+		planStatus:    motion.PlanStatus{State: motion.PlanStateSucceeded, Timestamp: time},
+	})
+}
+
+// a stateExecution is the struct held in the state that
+// holds the history of plans & plan status updates an
+// execution has exprienced & the waitGroup & cancelFunc
+// required to shut down an execution's goroutine.
+type stateExecution struct {
+	id            motion.ExecutionID
+	componentName resource.Name
+	waitGroup     *sync.WaitGroup
+	cancelFunc    context.CancelFunc
+	history       []motion.PlanWithStatus
+}
+
+func (e *stateExecution) stop() {
+	e.cancelFunc()
+	e.waitGroup.Wait()
+}
+
+var replanReason = "replanning"
+
+type componentStateByComponent map[resource.Name]componentState
+
+type componentState struct {
+	executionIDHistory []motion.ExecutionID
+	executionsByID     map[motion.ExecutionID]stateExecution
+}
+
+func (cs componentState) lastExecution() stateExecution {
+	return cs.executionsByID[cs.lastExecutionID()]
+}
+
+func (cs componentState) lastExecutionID() motion.ExecutionID {
+	return cs.executionIDHistory[0]
+}
+
+type newPlanMsg struct {
+	plan       motion.Plan
+	planStatus motion.PlanStatus
+}
+
+type stateUpdateMsg struct {
+	componentName resource.Name
+	executionID   motion.ExecutionID
+	planID        motion.PlanID
+	planStatus    motion.PlanStatus
+}
+
+// ReplanRequest is a temporary struct that triggers the execution to replan
+// if FailReason is non nil then the replanning will fail.
+type ReplanRequest struct {
+	FailReason *string
+}
+
+// ExecutionRequest is a temporary struct that triggers the end of an execution
+// if FailReason is non nil then the execution will fail.
+type ExecutionRequest struct {
+	// if nil then the execution succeeded
+	FailReason *string
+}
+
+// TestConfig is a temporary struct that holds channels for testing execution replanning, success & failure.
```

Comment:
Temporary in the sense that it's deprecated and you're planning on getting rid of it soon?

Reply:
Also, I don't think this needs to be public

Reply:
yes I plan on getting rid of this code soon. If it is a blocker for merging I can try to come up with a different approach for being able to tests the behavior when replanning, plan success & plan failure occur.